[PPP-3455] - Dynamic code injection vulnerabilities found

akhayrutdinov opened the pull request:
- in coreQueries.js, LegacyQuery.js, timeplot.js, replace `eval()` with `JSON.parse()`
- in json.js, add a function that involes `JSON.parse()` if it exists (should exist in modern browsers)
- in simile-ajax-bundle.js, copy changes from json.js and replace another `eval()` with `JSON.parse`

@diogofscmariano, @dcleao, @pamval, review it please.

pamval commented:
@jvelasques @afrjorge review please

akhayrutdinov commented:
@jvelasques, @afrjorge, review it please

akhayrutdinov commented:
@diogofscmariano, @dcleao, may be you guys can find some time for this PR?

diogofscmariano commented:
simile-ajax-bundle.js and timeplot.js cannot be updated because they are external libraries. In the same level, i think upgrading those is out of scope for now (i dont think they even have that)

diogofscmariano commented on core-js/src/main/javascript/cdf-legacy/queries/coreQueries.js:
```diff
@@ -681,7 +681,7 @@
       lastResultSet: {
         reader: function(json) {
-          json = eval("(" + json + ")");
+          json = JSON.parse(json);
```
For this, we might need some real unit tests to check if this is failing. For this, we must pick real query results and use them as base

diogofscmariano commented on core-js/src/main/javascript/cdf/queries/LegacyQuery.js:
```diff
@@ -53,7 +53,7 @@ define([
       lastResultSet: {
         reader: function(json) {
-          json = eval("(" + json + ")");
+          json = JSON.parse(json);
```
For this, we might need some real unit tests to check if this is failing. For this, we must pick real query results and use them as base

afrjorge commented:
@akhayrutdinov you need to update some unit tests, checkout this commit 705dd504e78b77e3a320b114f4acf8312098acb2 I think you might be able to cherry-pick it into your branch.

To build and run the unit tests execute:
 **mvn clean install**
Before that, you might need to edit the content of your **settings.xml** inside your **.m2** folder as described [here](https://github.com/webdetails/cdf).

akhayrutdinov commented:
@afrjorge, I'm getting test failures on the current master:
```
[INFO] Running 'karma start /home/akhayrutdinov/pentaho/cdf/core-js/src/test/javascript/config/karma.ci.conf.js --no-colors' in /home/akhayrutdinov/pentaho/cdf/core-js
       ERROR: '[error] WD: TypeError: 'undefined' is not an object (evaluating 'buttons[index].parentNode')'
       ERROR: '[error] WD: addComponent: duplicate component name 'PostInitMarker''
       ERROR: '[error] WD: addComponent: invalid component'
       ERROR: '[error] WD: addComponent: invalid component'
       ERROR: '[error] WD: addComponent: invalid component'
       ERROR: '[error] WD: addComponent: invalid component'
       ERROR: '[error] WD: No data available.'
       ERROR: '[error] WD: Error Message'
       PhantomJS 1.9.8 (Linux 0.0.0): Executed 205 of 208 SUCCESS (19.402 secs / 16.017 secs)
[ERROR] .write(string, encoding, offset, length) is deprecated. Use write(string[, offset[, length]][, encoding]) instead.
[INFO] 
[INFO] --- frontend-maven-plugin:0.0.26:karma (javascript-legacy-tests) @ cdf-core-js ---
[INFO] Running 'karma start /home/akhayrutdinov/pentaho/cdf/core-js/src/test/javascript/config/karma.ci.conf.legacy.js --no-colors' in /home/akhayrutdinov/pentaho/cdf/core-js
[INFO] WARN [web-server]: 404: /dummy/plugin/pentaho-cdf/api/storage/read?action=read&_=1455024661258
[INFO] PhantomJS 1.9.8 (Linux 0.0.0) ERROR: 'CDF: Object type AutoCompleteBoxComponent can't be mapped to a valid class'
[INFO] 
[INFO] PhantomJS 1.9.8 (Linux 0.0.0) ERROR: 'CDF: Object type dateInputComponent can't be mapped to a valid class'
[INFO] 
[INFO] PhantomJS 1.9.8 (Linux 0.0.0) ERROR: 'CDF: Object type radioComponent can't be mapped to a valid class'
[INFO] 
[INFO] PhantomJS 1.9.8 (Linux 0.0.0) ERROR: 'CDF: Object type checkComponent can't be mapped to a valid class'
[INFO] 
[INFO] PhantomJS 1.9.8 (Linux 0.0.0) ERROR: 'CDF: Object type SimpleAutoComplete can't be mapped to a valid class'
[INFO] 
[INFO] PhantomJS 1.9.8 (Linux 0.0.0) LOG: 'CDF: InitInstance 0'
[INFO] 
[INFO] PhantomJS 1.9.8 (Linux 0.0.0) ERROR: 'CDF: Object type unmanaged can't be mapped to a valid class'
[INFO] 
[INFO] WARN [web-server]: 404: /dummy/plugin/pentaho-cdf/api/comments/list?ts=1455024661308&action=list&page=generic&firstResult=0&maxResults=100&where=false
[INFO] WARN [web-server]: 404: /dummy/api/repos/%3Afake%3Aregions.xaction/generatedContent
       PhantomJS 1.9.8 (Linux 0.0.0): Executed 0 of 113 SUCCESS (0 secs / 0 secs)
       ERROR: 'CDF: Found error: [object Object] - error, Error: Not Found'
       ERROR: 'CDF: Error: Invalid XML: NOT FOUND'
       ERROR: 'CDF: Found error: [object Object] - error, Error: Not Found'
       ERROR: 'CDF: Error: Invalid XML: NOT FOUND'
[INFO] PhantomJS 1.9.8 (Linux 0.0.0): Executed 0 of 113 SUCCESS (0 secs / 0 secs)
       PhantomJS 1.9.8 (Linux 0.0.0): Executed 1 of 113 SUCCESS (0 secs / 0.117 secs)
       PhantomJS 1.9.8 (Linux 0.0.0): Executed 2 of 113 SUCCESS (0 secs / 0.225 secs)
       PhantomJS 1.9.8 (Linux 0.0.0): Executed 3 of 113 SUCCESS (0 secs / 0.329 secs)
       PhantomJS 1.9.8 (Linux 0.0.0): Executed 4 of 113 SUCCESS (0 secs / 0.433 secs)
       PhantomJS 1.9.8 (Linux 0.0.0): Executed 5 of 113 SUCCESS (0 secs / 0.537 secs)
       ERROR: 'CDF: Object type dummy can't be mapped to a valid class'
       ERROR: 'CDF: Object type dummy can't be mapped to a valid class'
       ERROR: 'CDF: Object type unmanaged can't be mapped to a valid class'
[INFO] PhantomJS 1.9.8 (Linux 0.0.0): Executed 64 of 113 SUCCESS (0 secs / 2.068 secs)
[INFO] WARN [web-server]: 404: /dummy/plugin/cda/api/doQuery?
       PhantomJS 1.9.8 (Linux 0.0.0) Dashboards.othersAwaitExecution behaviour unit testing # Should go through Dashboards.updateAll handling comp1priority5 and *only after comp1priority5's end* it should handle comp3priority10, when a call is made to comp3priority10 (due to its lower priority) FAILED
[INFO] 	Expected 0 to equal 1.
[INFO] 	Error: Expected 0 to equal 1.
[INFO] 	    at /home/akhayrutdinov/pentaho/cdf/core-js/src/test/javascript/cdf-legacy/paralel-query-calls-spec-legacy.js:288
[INFO] 	Expected 0 to equal 1.
[INFO] 	Error: Expected 0 to equal 1.
[INFO] 	    at /home/akhayrutdinov/pentaho/cdf/core-js/src/test/javascript/cdf-legacy/paralel-query-calls-spec-legacy.js:289
[INFO] 	Expected 0 to equal 1.
[INFO] 	Error: Expected 0 to equal 1.
[INFO] 	    at /home/akhayrutdinov/pentaho/cdf/core-js/src/test/javascript/cdf-legacy/paralel-query-calls-spec-legacy.js:290
       PhantomJS 1.9.8 (Linux 0.0.0) Dashboards.othersAwaitExecution behaviour unit testing # Should go through Dashboards.updateAll handling comp2priority5 and *not* comp1priority5 (as it is already under updating) and *only after comp2priority5's end* it should handle comp3priority10  (due to its lower priority), when a call is made to comp2priority5 FAILED
[INFO] 	Expected 1 to equal 0.
[INFO] 	Error: Expected 1 to equal 0.
[INFO] 	    at /home/akhayrutdinov/pentaho/cdf/core-js/src/test/javascript/cdf-legacy/paralel-query-calls-spec-legacy.js:354
[INFO] 	Expected 1 to equal 0.
[INFO] 	Error: Expected 1 to equal 0.
[INFO] 	    at /home/akhayrutdinov/pentaho/cdf/core-js/src/test/javascript/cdf-legacy/paralel-query-calls-spec-legacy.js:355
[INFO] 	Expected 1 to equal 0.
[INFO] 	Error: Expected 1 to equal 0.
[INFO] 	    at /home/akhayrutdinov/pentaho/cdf/core-js/src/test/javascript/cdf-legacy/paralel-query-calls-spec-legacy.js:356
       PhantomJS 1.9.8 (Linux 0.0.0): Executed 113 of 113 (2 FAILED) (5.849 secs / 5.785 secs)
```

Here are final results:
```
PhantomJS 1.9.8 (Linux 0.0.0): Executed 205 of 208 SUCCESS (19.402 secs / 16.017 secs)
PhantomJS 1.9.8 (Linux 0.0.0): Executed 113 of 113 (2 FAILED) (5.849 secs / 5.785 secs)
```

Do these test really fail or it is my environment's flaws?

akhayrutdinov commented:
Nevermind, it was likely an environment problem; `mvn clean` solved it.

I've updated the PR with tests and added a fall-back scenario if we suddenly get `'`. But normally this should not happen, because xactions use `JSONStringer` that respects proper string building (with `"`)

@diogofscmariano, @afrjorge, review it please

afrjorge commented on core-js/src/main/javascript/cdf-legacy/queries/coreQueries.js:
```diff
@@ -681,7 +681,15 @@
+            if (!(err instanceof SyntaxError)) {
+              throw err;
+            }
+            json = JSON.parse(str.replace(/'/g, '"'));
```
@akhayrutdinov @diogofscmariano In this case, because we might have a single quote in the content of a valid string, I think this will create a regression. As an example, change **LegacyQuery-spec.js** line **17**-**20** to:
`var unprocessedDataString = "{\"metadata\":[\"Sales\"],\"values\":[[\"Euro+ Shopping's Channel\",\"914.11\"],[\"Mini Gifts Ltd.\",\"6558.02\"]]}",`
`unprocessedData = {metadata: [{colIndex : 0, colType : "String", colName : "Name"},`
` {colIndex : 1, colType : "String", colName : "Sales"}],`
`resultset: [["Euro+ Shopping's Channel", "914.11"], ["Mini Gifts Ltd.", "6558.02"]]},`
Notice how the single quote in `"Euro+ Shopping's Channel"` is wrongly changed.

akhayrutdinov commented:
@afrjorge, after a small amendment in a *test* (replacing manual written JSON representation), all test pass w/o any fall back actions. Do I need to keep such?

I can offer to do the following:
```
var r = JSON.parse(original)

// caught SyntaxError
var tmp = original.replace(/'/g, '"')
try {
  // check whether the string is parse-able with replaced enclosures
  var t2 = JSON.parse(tmp);
catch (err) {
  // something wrong here
  throw err
}
// evaluate the original string: it is safe, because parsing competed successfully
r = eval(original)
```
That is quite hacky, but I suppose it suits the goal.

So, what do you think: should such kind of handling be done or not?

afrjorge commented:
@akhayrutdinov I think blindly replacing single quotes with double quotes is not correct, so I don't agree with using this:
```var tmp = original.replace(/'/g, '"')```
If we have server-side responses containing JSON that doesn't comply with the standard, an issue should be opened to tackle it server-side.
Your solution seems fine.

akhayrutdinov commented:
@afrjorge, can you please review it once again?
I added more test cases and error logging to raise attention if an parsing problem comes

afrjorge commented:
@akhayrutdinov I think you are making two wrong assumptions:
* the SyntaxError is exactly the single quote vs double quote surrounding strings issue
* all single quotes are ONLY used as string delimiters

For example, in LegacyQuery-spec.js using a single quote in the content of a string (```Euro+ Shopping's Channel```) fails (see 95786c64758a827a09e3658af38da67fabbc2edd in my branch [PPP-3455](https://github.com/afrjorge/cdf/tree/PPP-3455)).
I think your previous version was ok, I added an extra commit to update the unit tests, see 307173532042f6f1a0aac63ac33ceca1a99d0f44 in my branch [PPP-3455-ok](https://github.com/afrjorge/cdf/tree/PPP-3455-ok)

akhayrutdinov commented:
@afrjorge, ok, I've grabbed your branch and added one more test. Review it please :smile:
